boards: opta: external flash and BLE support

fogzot opened the pull request:
This set of changes adds support for QSPI-based external flash and Bluetooth to the device tree. This make it possible to correctly build and execute the fatfs and several Bluetooth samples out of the box.

Also added a function to read the external flash OTP to extract information about the Opta model and hardware features and a second function to retrieve the "official" Opta serial number.

fogzot committed:
boards: opta: external flash and BLE support

This set of changes adds support for QSPI-based external flash and
Bluetooth to the device tree. This make it possible to correctly build and
execute the fatfs and several Bluetooth samples out of the box.

Also added a function to read the external flash OTP to extract
information about the Opta model and hardware features and a second
function to retrieve the "official" Opta serial number.

Signed-off-by: Federico Di Gregorio <fog@dndg.it>

pillo79 commented on boards/arduino/opta/arduino_opta-common.dtsi:
Not sure it's OK to share the Flash controller description on both cores. Is accessing external Flash actually useful on the M4?

fogzot replied:
Probably it is OK to assign it by default to M7.

pillo79 commented on boards/arduino/opta/board_info.c:
```diff
@@ -0,0 +1,105 @@
+	HAL_StatusTypeDef hal_ret;
+
+	QSPI_CommandTypeDef cmd = {
+		.Instruction = 0x48,
```
Please define an `AT25SF128_READ_SECURITY_REGISTERS` and use this.

fogzot replied:
OK, but where to define it? Just in the Opta tree, in `board.h` or someplace more generic?

pillo79 reviewed (changes requested):
Thanks for this! There was actually another PR adding initial Flash support to Opta so I waited for that to get in before reviewing this one.
Please rebase and fix the conflicts, mostly LGTM except for the security read operation - to be accepted it should be reworked so it is implemented by the Flash driver.

fogzot commented on boards/arduino/opta/doc/index.rst:
```diff
@@ -80,6 +80,49 @@ as well as by the main PLL clock. By default, the CPU2 (Cortex-M4) System clock
+
+ .. code-block:: dts
+
+    partitions {
+        compatible = "fixed-partitions";
+        #address-cells = <1>;
+        #size-cells = <1>;
+
+        /* Partition 1: WiFi firmware and certificates 1MB */
+        wlan_partition: partition@1000 {
+            label = "wlan";
+            reg=<0x001000 DT_SIZE_K(1020)>;
+        };
+
+        /* Partition 2: OTA 5MB */
+        ota_partition: partition@100000 {
+            label = "ota";
+            reg=<0x100000 DT_SIZE_K(5 * 1024)>;
+        };
+
+        /* Partition 3: Provisioning KVStore 1MB */
+        kvs_partition: partition@600000 {
+            label = "kvs";
+            reg=<0x600000 DT_SIZE_K(1024)>;
+        };
+
+        /* Partition 4: User data / OPTA PLC runtime 7MB (littlefs) */
+        user_partition: partition@700000 {
+            label = "user";
+            reg=<0x00000 DT_SIZE_K(7 * 1024)>;
+        };
+    };
```
OK.

fogzot commented on boards/arduino/opta/board_info.c:
```diff
@@ -0,0 +1,105 @@
+
+const struct device *const dev = DEVICE_DT_GET(DT_NODELABEL(external_flash));
+
+struct flash_stm32_qspi_data {
+	QSPI_HandleTypeDef hqspi;
+	struct k_sem sem;
+	/* More fields here but we're not interested in them. */
+};
+
+static int board_info(void)
+{
+	struct flash_stm32_qspi_data *dev_data = dev->data;
+
```
OK, but at which level should the operation be defined? I'd say that the only way for it to work is to add it to `flash_stm32_ex_op.c` but then where should the AT25SF128_READ_SECURITY_REGISTERS be defined? The AT25SF128 is not st-specific. Can you tell me where do you like the code to be placed?

pillo79 reviewed (commented):
> I see some small differences between the current `quadspi` definition and mine:
> 
>   1. bus width is set to 2 but the ST quadspi supports 4 without problems;
> 
>   2. maximum frequency is 70 MHz while the datasheet specifies 80 MHz.
> 
> It is OK to keep my values or there is indication that Zephyr needs the current values?

Yes, please use your values - Opta should have no issues supporting 4 lanes and 80MHz clocks.

fogzot commented:
I just pushed a new version that builds over the additions of "ex op" to the STM32 QSPI flash driver. Probably this is still not perfect but it shows how the request to not use HAL directly can be addressed.

fogzot commented:
The "generic ex op" for STM32 QSPI flash got in, so I rebased this patch on it and changed the code to use the extended operation. @pillo79, let me know if anything else needs changes.

pillo79 reviewed (commented):
Looks good, still a couple of minor fixes.

etienne-lms reviewed (commented):
Some minor comments, otherwise LGTM.

fogzot commented:
@etienne-lms I just pushed all requested changes.

etienne-lms commented on boards/arduino/opta/board_info.c:
```diff
@@ -0,0 +1,84 @@
+		return -ENODEV;
+	}
+
+	int ret = flash_ex_op(dev, FLASH_STM32_QSPI_EX_OP_GENERIC_READ, (uintptr_t)&cmd,
+			      &opta_board_info);
```
Nitpicking suggestion:
```c
	return flash_ex_op(dev, FLASH_STM32_QSPI_EX_OP_GENERIC_READ, (uintptr_t)&cmd,
			   &opta_board_info);

etienne-lms reviewed (dismissed):
LGTM with comment addressed or not.

fogzot commented on boards/arduino/opta/arduino_opta-external-flash-partitioning.dtsi:
```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2024 DNDG srl
```
Some files are from 2024, but I can just put 2025 if it is better.

fogzot commented on samples/subsys/fs/fs_sample/boards/arduino_opta_stm32h747xx_m7.conf:
```diff
@@ -0,0 +1,14 @@
+# SPDX-License-Identifier: Apache-2.0
+#
+
+# Usually some files are already present of the first partition,
+# don't create new ones or format partition in case of failure.
```
Done.

GeorgeCGV commented on boards/arduino/opta/arduino_opta_stm32h747xx_m7.dts:
```diff
@@ -123,33 +124,84 @@ zephyr_udc0: &usbotg_fs {
 
-	at25sf128a: qspi-nor-flash@0 {
+	qspi_flash: qspi-nor-flash@90000000 {
 		compatible = "st,stm32-qspi-nor";
-		reg = <0>;
+		reg = <0x90000000 DT_SIZE_M(16)>;
```
Needs to account for #89942

pillo79 commented on boards/arduino/opta/board_info.c:
```diff
@@ -0,0 +1,84 @@
+#define AT25SF128_READ_SECURITY_REGISTERS 0x48
+
+static struct opta_board_info opta_board_info = {0};
+static char serial_number[OPTA_SERIAL_NUMBER_SIZE + 1] = {0};
```
This was fact actually useful for the serial number, since the very first call to `opta_get_serial_number` is expecting to find a zero-filled buffer (and return a zero-terminated string):

https://github.com/zephyrproject-rtos/zephyr/blob/51bde72671d2124fe18cbe10fd95705c25a9e16d/boards/arduino/opta/board_info.c#L78-L83

Please restore this to ensure proper operation, otherwise LGTM!

fogzot replied:
Reading @etienne-lms comment I tought that statics where automatically zeroed. If they are not then the initialization is necessary. I'll revert the change.

pillo79 replied:
Yep, my bad, forgot that actually are zero initialized already. Sorry for the misdirection! 🤦‍♂️
